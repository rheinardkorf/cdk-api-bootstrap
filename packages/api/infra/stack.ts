import * as cdk from '@aws-cdk/core';
import * as apigateway from '@aws-cdk/aws-apigateway';
import * as dynamodb from '@aws-cdk/aws-dynamodb';
import * as lambda from '@aws-cdk/aws-lambda';
import * as yaml from 'yamljs';
import * as path from 'path';
import * as fs from 'fs';

const wsConfig = JSON.parse(fs.readFileSync(path.resolve(__dirname, '../../../workspace.config.json')).toString());
const stage = process.env.STAGE || "dev";
const stackId = `${wsConfig.stackPrefix}API-${stage}`;
const tableName = `${wsConfig.defaultTable}-${stage}`;

export class ServiceStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Get reference to DynamoDB table.
    // Must specify `globalIndexes` for permissions to work on indexes.
    const table = dynamodb.Table.fromTableAttributes(this,`${stackId}_DynamoDB`, {
        tableName,
        globalIndexes: ['GSI1', 'GSI2', 'GSI3']
    });

    // Create API Gateway.
    const api = new apigateway.RestApi(this, `${stackId}_API`, {
      defaultCorsPreflightOptions: {
        allowOrigins: apigateway.Cors.ALL_ORIGINS,
        allowHeaders:['Content-Type', 'X-Amz-Date', 'X-Amz-Security-Token', 'Authorization', 'X-Api-Key', 'X-Requested-With', 'Accept', 'Access-Control-Allow-Methods', 'Access-Control-Allow-Origin', 'Access-Control-Allow-Headers']
      }
    });

    // Create common layer for include dependencies in all lambdas.
    const commonLayer = new lambda.LayerVersion(this, `${stackId}_Common`, {
        code: new lambda.AssetCode(`layers/common`),
        compatibleRuntimes: [lambda.Runtime.NODEJS_14_X]
    })

    // Custom Authorizer Function.
    const lambdaAuthorizerFunction = new lambda.Function(this, `${stackId}_Authorization_Handler`, {
      runtime: lambda.Runtime.NODEJS_14_X,
      code: new lambda.AssetCode(`authorizer`),
      handler: `authorizer.authorizer`,
      layers: [commonLayer]
    })

    // Custom Authorizer.
    const lambdaAuthorizer = new apigateway.TokenAuthorizer(this, `${stackId}_TokenAuthorizer`, {
      handler: lambdaAuthorizerFunction
    })

    // Load the API Spec.
    const spec = yaml.load(path.resolve(__dirname, '../api-spec.yaml'))

    // Generate resources for API paths and methods.
    for ( const key in spec.paths ) {

      // New API Gateway resource for path.
      const resource = api.root.resourceForPath(key);

      // Get the root of the path to point to functions.
      const fullPath = key.replace(/^\//,'')
      const root = fullPath.split('/')[0]

      // For each of the path methods add a new API Gateway method.
      for ( const method in spec.paths[key]) {
        const pathMethod = spec.paths[key][method]
        const isSecured = pathMethod.security && true || false

        // Create a new Lambda Function to handle the method.
        const lambdaHandler = new lambda.Function(this, `${pathMethod.operationId}_Handler`, {
          runtime: lambda.Runtime.NODEJS_14_X,
          code: new lambda.AssetCode(`functions/${root}`),
          handler: `${pathMethod.operationId}.${pathMethod.operationId}`,
					layers: [commonLayer],
          environment: {
            TABLE_NAME: table.tableName
          },
          
        })

        if ( 'GET' === method.toUpperCase() ) {
          table.grantReadData(lambdaHandler);
        } else {
          table.grantReadWriteData(lambdaHandler);
        }

        const authorizer = isSecured && {
          authorizationType: apigateway.AuthorizationType.CUSTOM,
          authorizer: lambdaAuthorizer
        } || {};

        // Attach the Lambda function to the API Gateway resource.
        resource.addMethod(method.toUpperCase(), new apigateway.LambdaIntegration(lambdaHandler), {
          ...authorizer
        });
      }
    }
  }
}