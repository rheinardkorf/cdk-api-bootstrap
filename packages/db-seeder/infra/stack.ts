import * as cdk from '@aws-cdk/core';
import * as dynamodb from '@aws-cdk/aws-dynamodb';
import * as lambda from '@aws-cdk/aws-lambda';
import * as cr from '@aws-cdk/custom-resources';
import * as logs from '@aws-cdk/aws-logs';
import * as iam from '@aws-cdk/aws-iam';
import * as path from 'path'
import * as fs from 'fs'

const wsConfig = JSON.parse(fs.readFileSync(path.resolve(__dirname, '../../../workspace.config.json')).toString());
const stage = process.env.STAGE || "dev";
const stackId = `${wsConfig.stackPrefix}DB-Seeder-${stage}`;
const tableName = `${wsConfig.defaultTable}-${stage}`;
const seedsStage = process.env.SEEDS_STAGE || stage;

export class SeederStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const table: dynamodb.ITable = dynamodb.Table.fromTableAttributes(this,`${stackId}__DynamoDB`, {
        tableName,
        globalIndexes: ['GSI1', 'GSI2', 'GSI3']
    });

    const seederResourceRole = new iam.Role(this,`${stackId}_seederResourceRole`,{
        assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com')
    })

    seederResourceRole.addToPolicy(new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        resources: ['*'],
        actions: [
            "dynamodb:GetItem",
            "dynamodb:BatchGetItem",
            "dynamodb:Query",
            "dynamodb:PutItem",
            "dynamodb:UpdateItem",
            "dynamodb:DeleteItem",
            "dynamodb:BatchWriteItem",
            "dynamodb:updateTable",
            "dynamodb:describeTable"
        ]
    }));

    const seedLayer = new lambda.LayerVersion(this, `${stackId}_SeedLayer`, {
        code: new lambda.AssetCode(`seeds`)
    })

    table.grantReadWriteData(seederResourceRole);
    const onEvent = new lambda.Function(this, `${stackId}-${tableName}-seedHandler`, {
        runtime: lambda.Runtime.NODEJS_14_X,
        handler: 'seeder.handler',
        code: lambda.Code.fromAsset(path.resolve(__dirname, '../seeder')),
        environment: {
            SEED_STAGE: seedsStage,
            TABLE_NAME: tableName
        },
        layers: [seedLayer],
        timeout: cdk.Duration.minutes(15),
        role: seederResourceRole
    })
    table.grantReadWriteData(onEvent);

    const seederProvider = new cr.Provider(this, `${stackId}_seederProvider`, {
        onEventHandler: onEvent,
        logRetention: logs.RetentionDays.ONE_DAY
    })

    /**
     * NOTE: For some reason the custom resource is not triggering
     * the "Update" action for Custom Resources. You will need
     * to destroy and deploy for changes in seed data to be reflected.
     */
    new cdk.CustomResource(this, `${stackId}_seederCustomResource`, {
        serviceToken: seederProvider.serviceToken
    })

  }
}