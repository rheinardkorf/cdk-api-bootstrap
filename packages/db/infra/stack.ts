import * as cdk from "@aws-cdk/core";
import * as dynamodb from "@aws-cdk/aws-dynamodb";
import * as path from 'path'
import * as fs from 'fs'

const wsConfig = JSON.parse(fs.readFileSync(path.resolve(__dirname, '../../../workspace.config.json')).toString());
const stage = process.env.STAGE || "dev";
const stackId = `${wsConfig.stackPrefix}DB-${stage}`;
const tableName = `${wsConfig.defaultTable}-${stage}`;

export class ServiceStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);
    
    const removalPolicy = !["prod", "production"].includes(stage)
      ? cdk.RemovalPolicy.DESTROY
      : cdk.RemovalPolicy.RETAIN;

    // DynamoDB Table Definition.
    const table = new dynamodb.Table(this, `${stackId}_DynamoDB`, {
      partitionKey: { name: "PK", type: dynamodb.AttributeType.STRING },
      sortKey: { name: "SK", type: dynamodb.AttributeType.STRING },
      billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
      tableName,
      removalPolicy,
      stream: dynamodb.StreamViewType.NEW_AND_OLD_IMAGES,
      pointInTimeRecovery: true
    });

    // Add GSI1.
    table.addGlobalSecondaryIndex({
      indexName: "GSI1",
      partitionKey: { name: "PK1", type: dynamodb.AttributeType.STRING },
      sortKey: { name: "SK1", type: dynamodb.AttributeType.STRING },
      projectionType: dynamodb.ProjectionType.ALL,
    });

    // Add GSI2.
    table.addGlobalSecondaryIndex({
      indexName: "GSI2",
      partitionKey: { name: "PK2", type: dynamodb.AttributeType.STRING },
      sortKey: { name: "SK2", type: dynamodb.AttributeType.STRING },
      projectionType: dynamodb.ProjectionType.ALL,
    });

    // Add GSI3.
    table.addGlobalSecondaryIndex({
      indexName: "GSI3",
      partitionKey: { name: "PK3", type: dynamodb.AttributeType.STRING },
      sortKey: { name: "SK3", type: dynamodb.AttributeType.STRING },
      projectionType: dynamodb.ProjectionType.ALL,
    });
  }
}
