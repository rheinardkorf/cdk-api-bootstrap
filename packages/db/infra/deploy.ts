#!/usr/bin/env node
import 'source-map-support/register';
import * as cdk from '@aws-cdk/core';
import { ServiceStack } from './stack';
import * as path from 'path';
import * as fs from 'fs';

const wsConfig = JSON.parse(fs.readFileSync(path.resolve(__dirname, '../../../workspace.config.json')).toString());
const stage = process.env.STAGE || 'dev';
const stackId = `${wsConfig.stackPrefix}DB-${stage}`;
const account = process.env.CDK_DEFAULT_ACCOUNT;
const region = process.env.CDK_DEFAULT_REGION || 'ap-southeast-2';

const app = new cdk.App();
new ServiceStack(app, stackId, {
  env: {
    account,
    region,
  },
  description: `Deploying Stack (${stackId})`
});