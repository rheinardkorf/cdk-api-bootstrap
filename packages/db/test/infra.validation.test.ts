import { SynthUtils } from "@aws-cdk/assert"
import { Stack } from "@aws-cdk/core"

import { ServiceStack } from "../infra/stack"

test("DB stack validation", () => {
    const stack = new Stack();
    const service_stack = new ServiceStack(stack, 'service_stack');
    const received = SynthUtils.toCloudFormation(service_stack);
    expect(SynthUtils.toCloudFormation(service_stack)).toMatchSnapshot();
});