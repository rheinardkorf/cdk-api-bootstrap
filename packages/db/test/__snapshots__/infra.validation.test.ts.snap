// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DB stack validation 1`] = `
Object {
  "Resources": Object {
    "MyProjectDBdevDynamoDBFC24880B": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "PK",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "SK",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "PK1",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "SK1",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "PK2",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "SK2",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "PK3",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "SK3",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "GlobalSecondaryIndexes": Array [
          Object {
            "IndexName": "GSI1",
            "KeySchema": Array [
              Object {
                "AttributeName": "PK1",
                "KeyType": "HASH",
              },
              Object {
                "AttributeName": "SK1",
                "KeyType": "RANGE",
              },
            ],
            "Projection": Object {
              "ProjectionType": "ALL",
            },
          },
          Object {
            "IndexName": "GSI2",
            "KeySchema": Array [
              Object {
                "AttributeName": "PK2",
                "KeyType": "HASH",
              },
              Object {
                "AttributeName": "SK2",
                "KeyType": "RANGE",
              },
            ],
            "Projection": Object {
              "ProjectionType": "ALL",
            },
          },
          Object {
            "IndexName": "GSI3",
            "KeySchema": Array [
              Object {
                "AttributeName": "PK3",
                "KeyType": "HASH",
              },
              Object {
                "AttributeName": "SK3",
                "KeyType": "RANGE",
              },
            ],
            "Projection": Object {
              "ProjectionType": "ALL",
            },
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "PK",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "SK",
            "KeyType": "RANGE",
          },
        ],
        "PointInTimeRecoverySpecification": Object {
          "PointInTimeRecoveryEnabled": true,
        },
        "StreamSpecification": Object {
          "StreamViewType": "NEW_AND_OLD_IMAGES",
        },
        "TableName": "data-dev",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
